### PR TITLE
ASoC: SOF: handle unaligned read position in debugfs reads

### LINKED ISSUE #792: Upstream: v5:  debugfs alignment issue?
Feedback from Takashi Iwai:

````
> +static ssize_t sof_dfsentry_read(struct file *file, char __user *buffer,
> +				 size_t count, loff_t *ppos)
> +{
> +	struct snd_sof_dfsentry *dfse = file->private_data;
> +	struct snd_sof_dev *sdev = dfse->sdev;
> +	int size;
> +	u32 *buf;
> +	loff_t pos = *ppos;
> +	size_t size_ret;
> +
> +	size = dfse->size;
> +
> +	/* validate position & count */
> +	if (pos < 0)
> +		return -EINVAL;
> +	if (pos >= size || !count)
> +		return 0;
> +	/* find the minimum. min() is not used since it adds sparse warnings */
> +	if (count > size - pos)
> +		count = size - pos;
> +
> +	/* intermediate buffer size must be u32 multiple */
> +	size = ALIGN(count, 4);

Doesn't pos need to be aligned to 32bit as well (at least at actually
reading from iomem)?
````

### PR DESCRIPTION
**Pull request by kv2019i**
Read position for debugfs reads may not be aligned to 32bit and
this may lead to invalid accesses to iomem. Handle unaligned
positions explicitly.

Feedback from Takashi Iwai

Fixes #792 

Signed-off-by: Kai Vehmanen <kai.vehmanen@linux.intel.com>

### CODE REVIEW

**Comment by kv2019i** [comment]
SOFCI TEST

**Comment by kv2019i** [comment]
Issues with CI still, the most recent is a warning now treated as error. Talked with @xiulipan and he was working on this. i tried to retrigger without a commit change, but didn't work, I'll try with a foobar commit.

**Comment by kv2019i** [comment]
>  i tried to retrigger without a commit change, but didn't work, I'll try with a foobar commit.

Hmm, even that didn't retrigger CI. :P Not sure what to try next.

**Comment by xiulipan** [comment]
@kv2019i 
The sof-ci is already workaround the machine driver build.
Travis-ci is controlled by some in-repo scripts, I could do nothing to it now.

**Comment by ranj063** [comment]
> Looks good to me, I didn't find any issues with the pointer and buffer management
> @ranj063 can you take a look?

@plbossart I just discussed a minor issue in the patch with the size going over dfse->size. @kv2019i is going to take care of it.

**Comment by kv2019i** [comment]
@ranj063 wrote:
> @plbossart I just discussed a minor issue in the patch with the size going over dfse->size. @kv2019i is going to take care of it.

Ack, working on a patch, will upload shortly. I made a quick test case and indeed the trace client gets killed if I try to do a non-aligned read that would go past dfse size. Good catch @ranj063 !

**Comment by kv2019i** [comment]
> Ack, working on a patch, will upload shortly. 

Done, now passes a test with unaligned read past the iomem region end.

File: sound/soc/sof/debug.c [hunk 1]
```diff
@@ -38,9 +39,20 @@ static ssize_t sof_dfsentry_read(struct file *file, char __user *buffer,
...
+
 	/* intermediate buffer size must be u32 multiple */
 	size = ALIGN(count, 4);
 
+	/* if start position is unaligned, read extra u32 */
+	if (unlikely(pos != *ppos)) {
+		size += 4;
+		skip = *ppos - pos;
+		if (pos + size > dfse->size)
+			size -= 4;
```
Comment (plbossart) [inline sound/soc/sof/debug.c]: could we instead only increment size when possible, e.g

````c
if (unlikely(pos != *ppos)) {
		skip = *ppos - pos;
		if (pos + size +4  <= dfse->size)
			size += 4;
````

the +4 then -4 looks odd.

**Comment by lyakh** [comment]
my approval is subject to addressing the above comment from @plbossart

File: sound/soc/sof/debug.c (hunk 1 above)
Comment (kv2019i) [reply sound/soc/sof/debug.c]: Ack. Yes, this is better, fixed and reuploaded.

**Pull request merged**